bw2data/project.py

```diff
@@ -128,10 +129,41 @@ def apply_revision(self, revision: dict) -> None:
         """
         from bw2data import revisions
 
+        # Once *any* revision is applied, we need to track changes
+        # Otherwise we could apply a patch, then make an untracked change, and we wouldn't know if
+        # a second patch could be safely applied in the future.
+        if not self.is_sourced:
+            self.is_sourced = True
+
+        # Patch parent can be a revision ID or `None`
+        # Local state can be a revision ID or `None`
+        # There are five possible cases:
+        # 1. patch parent: None | local: None
+        # Beginning of event sourcing, applying patch leads to new local state
+        # No real guarantee that they start from the same state, but we can hope for the best
+        # 2. patch parent: int | local: None
+        # Can't apply safely as patch state could be different than local
+        # Raise `PossibleInconsistentData`
+        # 3. patch parent: None | local: int
+        # Can't apply safely as local state could be different than patch
+        # Raise `PossibleInconsistentData`
+        # 4. patch parent: int | local: int | patch parent == local
+        # Can apply safely, starting state is consistent
+        # 5. patch parent: int | local: int | patch parent != local
+        # Raise `InconsistentData`
         meta = revision["metadata"]
-        parent = meta.get("parent_revision")
-        assert not parent or self.revision == parent
-        assert parent or not self.revision
+        local, remote = self.revision, meta.get("parent_revision")
+        if local is remote is None:
+            pass
+        elif local and not remote:
+            raise PossibleInconsistentData
+        elif not local and remote:
+            raise PossibleInconsistentData
+        elif local and remote and local == remote:
+            pass
+        elif local and remote and local != remote:
+            raise InconsistentData
```

Comment:
This would be a good candidate for a separate function which could be reused and tested, but that can be left for a future enhancement.